### PR TITLE
[sc-34030]:  Update QML CI to be able to build JSON output. De-duplicate build-pr.

### PR DESCRIPTION
**Pull request by rashidnhm**
**Title:** Update QML CI to be able to build JSON output. De-duplicate build-pr.

**Summary:**
This PR updates the QML CI Workflow to be able to build sphinx with JSON serialization. Note that there is no workflow being added to actually build the docs as JSON (that is coming soon), but this PR adds the capability of being able to do so.

It also de-duplicates build-pr.yml by having it use the new build-branch re-usable workflow.

**Relevant references:**
[sc-34030](https://app.shortcut.com/xanaduai/story/34030/update-qml-sphinx-build-workflow-to-output-demo-files-as-json-to-be-ingested-by-the-new-pennylane-site)

**Possible Drawbacks:**
build-pr can exhibit unexpected behaviour but if anything is noticed it can be patched.

**Related GitHub Issues:**
None.

### CODE REVIEW

File: .github/workflows/qml_pipeline_utils/qml_pipeline_utils/services/parse_execution_times.py [hunk 1]
```diff
@@ -77,12 +89,20 @@ def parse_execution_times(num_workers: int, offset: int,
         sphinx_build_directory: The directory where sphinx outputs the built demo html files
         sphinx_gallery_dir_name: The gallery directory name inside sphinx_build_directory
                                  where sphinx puts all gallery demo html files
+        sphinx_build_type: The output format of sphinx-build, Valid values are "html" and "json"
         glob_pattern: The pattern use to glob all demonstration files inside build_directory. Defaults to "*.py"
     """
 
+    assert sphinx_build_type in {"html", "json", "fjson"}, "Invalid sphinx build type"
```
Comment (sanchitbapat) [inline .github/workflows/qml_pipeline_utils/qml_pipeline_utils/services/parse_execution_times.py:96]: Does it make sense to add `fjson` to the [assertion above](https://github.com/PennyLaneAI/qml/pull/715/files#diff-b51f73e3deaf542b434a71b6179d6f27f5e07774a9c57dcfcf4008fcb07f2418R52)?

File: .github/workflows/build-branch.yml [hunk 2]
```diff
@@ -272,6 +289,7 @@ jobs:
           path: |
             _build/html
             !_build/html/*.html
+            !_build/html/*.fjson
```
Comment (sanchitbapat) [inline .github/workflows/build-branch.yml:292]: Curious, does it also generate `.json` files and should we upload those?

**Review by sanchitbapat — APPROVED** [review]
LGTM! As long as the PR builds works 😄 
Just curious when would `sphinx_build_type=json` be used?

File: .github/workflows/build-branch.yml (hunk 2 above)
Comment (rashidnhm) [reply .github/workflows/build-branch.yml:292]: It does not, sphinx outputs all files with the .fjson suffix (not sure why though)

Comment (rashidnhm) [reply .github/workflows/build-branch.yml:292]: See: https://www.sphinx-doc.org/en/master/usage/builders/index.html#sphinxcontrib.serializinghtml.JSONHTMLBuilder

File: .github/workflows/qml_pipeline_utils/qml_pipeline_utils/services/parse_execution_times.py (hunk 1 above)
Comment (rashidnhm) [reply .github/workflows/qml_pipeline_utils/qml_pipeline_utils/services/parse_execution_times.py]: I had a quick think about this, and it doesn't make sense to support it on the Python side either, as you'd either be building with html or json serializer. I've removed the fjson support from the python side and now it is json/html across the board.

**Comment by rashidnhm** [comment]
> LGTM! As long as the PR builds works smile Just curious when would `sphinx_build_type=json` be used?

Great question @sanchitbapat, this is pre-requisite work for the next phase release of the new pennylane.ai react site. Currently we have Sphinx build the entire QML website that we serve to users, but since we want react to build it, it needs to ingest the content of the demos somehow. This capability of sphinx allows the build of just the demo (in html format) that is then encapsulated within a .fjson file. No styling or css / mathjax JS is added. The react build pipeline will ingest these files when building demo pages on that end.

**Comment by sanchitbapat** [comment]
> 

Got it, thanks!

**Pull request merged**